tee/src/lib.rs

```diff
@@ -0,0 +1,669 @@
+		ClusterIsAlreadyFull,
+		/// Enclave already assigned to a cluster
+		EnclaveAlreadyAssigned,
+		/// Enclave not assigned to a cluster
+		EnclaveNotAssigned,
+		/// Enclave does not exists
+		EnclaveDoesNotExists,
+		/// Internal logical error
+		InternalLogicalError,
+		/// Unknown enclave operator account
+		UnknownEnclaveOperatorAccount,
+		/// Error Adding To Queue
+		ErrorAddingToQueue,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::register_enclave())]
+		pub fn register_enclave(
+			origin: OriginFor<T>,
+			enclave_address: Vec<u8>,
+			api_uri: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let account = ensure_signed(origin)?;
+
+			ensure!(api_uri.len() < T::MaxUriLen::get().into(), Error::<T>::UriTooLong);
+			ensure!(api_uri.len() > T::MinUriLen::get().into(), Error::<T>::UriTooShort);
+
+			let result = AccountEnclaveId::<T>::get(&account)
+				.and_then(|enclave_id| EnclaveData::<T>::get(enclave_id).map(|enc| enc));
+
+			if result.is_some() {
+				let enclave = result.ok_or(Error::<T>::EnclaveDoesNotExists)?;
+				ensure!(
+					enclave.enclave_address != enclave_address,
+					Error::<T>::EnclaveAddressAlreadyRegisteredtoTheAccount
+				)
+			}
+
+			let (enclave_id, new_id) = Self::new_enclave_id()?;
+			// Needs to have enough money
+			let imbalance = T::Currency::withdraw(
+				&account,
+				T::EnclaveFee::get(),
+				WithdrawReasons::FEE,
+				KeepAlive,
+			)?;
+			T::FeesCollector::on_unbalanced(imbalance);
+
+			let enclave = Enclave::new(api_uri.clone(), enclave_address);
+
+			AccountEnclaveId::<T>::insert(account.clone(), enclave_id);
+			EnclaveData::<T>::insert(enclave_id, enclave);
+			EnclaveIdGenerator::<T>::put(new_id);
+
+			let reg_enclaves: BoundedVec<EnclaveId, T::MaxRegisteredEnclaves> =
+				<EnclaveRegistrationList<T>>::get();
+
+			let reg_enclaves = reg_enclaves
+				.try_mutate(|v| v.push(enclave_id))
+				.ok_or(Error::<T>::ErrorAddingToQueue)?;
+			<EnclaveRegistrationList<T>>::put(reg_enclaves);
+
+			Self::deposit_event(Event::EnclaveAdded { account, api_uri, enclave_id });
+			Ok(().into())
+		}
+
+		/// Removes an enclave from the system
+		/// Origin- operator account address
+		/// If the enclave is assigned, it will be placed in queue for tech committee approval
+		/// If enclave is not already assigned, he can exit without permission.
+		#[pallet::weight(T::WeightInfo::unregister_enclave())]
+		pub fn unregister_enclave(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let account = ensure_signed_or_root(origin)?;
+			match account {
+				Some(account_id) => {
+					let enclave_id = AccountEnclaveId::<T>::get(account_id.clone())
+						.ok_or(Error::<T>::UnknownEnclaveOperatorAccount)?;
+
+					let unregistered_enclaves: BoundedVec<EnclaveId, T::MaxUnRegisteredEnclaves> =
+						<EnclaveUnregistrationList<T>>::get();
+
+					let reg_enclaves = <EnclaveRegistrationList<T>>::get();
+
+					if !reg_enclaves.contains(&enclave_id) {
```

Comment:
Here it should not work like that :

You ask for registration -> You're put in the registration list -> When you're assigned, the registration list is cleaned. 

If you are in the registration list, it means you're not assigned yet.

Also there should be a root call to just remove a registration from the registration list. (Something called remove_registration)